R/render-table.R

```diff
@@ -37,23 +104,96 @@ renderTable <- function(expr, ..., env=parent.frame(), quoted=FALSE, func=NULL)
     xtable_args <- dots[intersect(names(dots), xtable_argnames)]
     non_xtable_args <- dots[setdiff(names(dots), xtable_argnames)]
 
-    # Call xtable with its args
+    # A small helper function to determine if the row.names can be coerced
+    # to numeric or if they are legitimate strings
+    isNumber <- function(n) {
+      grepl("^\\d$+", n)
+    }
+
+    # Figure out column alignment
+    ## Case 1: if align=NULL, check if row.names are numbers. If not, make
+    ## sure to left-align them (xtable right-aligns them by default, which
+    ## looks weird when the row.names are strings).
+    if (is.null(align)) {
+      cols <- if (isNumber(row.names(data))) "r" else "l"
+      for (i in seq_len(ncol(data))) {
+        cls <- class(data[,i])
+        if (cls == "numeric" || cls == "integer") cols <- paste0(cols, "r")
+        else cols <- paste0(cols, "l")
+      }
+    } else {
+      ## Case 2: if align!=NULL, check if it is only one character or a vector
+      ## and process it accordingly.
+      num_cols <- if (rownames) nchar(align) else nchar(align)+1
+      valid <- !grepl("[^lcr]", align)
+      if (num_cols == ncol(data)+1 && valid) {
+        cols <- if (rownames) align else paste0("r", align)
+      } else if (nchar(align) == 1 && valid) {
+        cols <- paste0(rep(align, ncol(data)+1), collapse="")
+      } else {
+        stop("`align` must contain only the characters `l`, `c` and/or `r` and have
+             length either equal to 1 or to the total number of columns")
+      }
+    }
+
+    # Call xtable with its (updated) args
+    xtable_args <- c(xtable_args, align=cols, digits=digits)
     xtable_res <- do.call(xtable, c(list(data), xtable_args))
 
     # Set up print args
     print_args <- list(
       xtable_res,
-      type = 'html',
-      html.table.attributes = paste('class="', htmlEscape(classNames, TRUE),
-                                    '"', sep='')
-    )
+      type='html',
+      include.rownames=rownames,
+      include.colnames=colnames,
+      NA.string=na,
+      html.table.attributes=paste0("class='", htmlEscape(classNames, TRUE), "' ",
+                                   "style='width:", validateCssUnit(width),";'"))
+
     print_args <- c(print_args, non_xtable_args)
 
-    return(paste(
+    # Capture the raw html table returned by print.xtable(), and store it in
+    # a variable for further processing
+    tab <- paste(
       utils::capture.output(
         do.call(print, print_args)
       ),
       collapse="\n"
-    ))
+    )
+
+    # All further processing concerns the table headers, so we don't need to run
+    # any of this if colnames=FALSE
+    if (colnames) {
+      # Make sure that the final html table has a proper header (not included
+      # in the print.xtable() default)
+      tab <- sub("<tr>", "<thead> <tr>", tab)
+      tab <- sub("</tr>", "</tr> </thead> <tbody>", tab)
+      tab <- sub("</table>", "</tbody> </table>", tab)
```

Comment:
@wch I'm not sure I understand what you're suggesting specifically... @jcheng5 Following your suggestion, I changed `"</table>"` to `"</table>$"`, but I think the second command (line 170) still falls under the naive territory (I'm not worried about the previous line because it's impossible for the user to enter anything before the first `"<tr>"`). However I suppose it is theoretically possible for one of the headers to be the string `"</tr>"`. How do we fix this?